### PR TITLE
add: company authentication page

### PR DESCRIPTION
**Pull request by akaJSaunders**
### JIRA link

https://companieshouse.atlassian.net/browse/LFA-1266

### Change description

Feature flag included

Test Company Number for Docker enviroment: SC109400 (use any legacy penalty reference)

Authorization code is 222222

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/ContainerFactory.ts [hunk 1]
```diff
@@ -78,6 +81,27 @@ export function createContainer(): Container {
...
 
+    const companyAuthConfig: CompanyAuthConfig = {
+        oath_scope_prefix: getEnvOrThrow('OATH_SCOPE_PREFIX'),
+        accountUrl: getEnvOrThrow('ACCOUNT_URL'),
+        accountRequestKey: getEnvOrThrow('OAUTH2_REQUEST_KEY'),
+        accountClientId: getEnvOrThrow('OAUTH2_CLIENT_ID'),
+        chsUrl: getEnvOrThrow('ACCOUNT_WEB_URL'),
+    };
+
+    const sessionStoreForAuthConfig = {
```
Comment (geek-soft) [inline src/ContainerFactory.ts:92]: I suggest creating class for this config which has static method `createFromEnvironmentVariables()` which returns instance populated with values read from env variables. This way it can be used here and in base controller to remove duplicates and keeps defaults consistent.

Example usage:

```
const config = SessionConfig.createFromEnvironmentVariables();
```

File: src/models/CompanyAuthConfig.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export default interface CompanyAuthConfig {
```
Comment (geek-soft) [inline src/models/CompanyAuthConfig.ts:1]: Why???

No default exports please.

File: src/models/sessionStoreConfig.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export default interface SessionStoreConfig {
```
Comment (geek-soft) [inline src/models/sessionStoreConfig.ts:1]: Why???

No default exports please.

File: test/ApplicationFactory.ts [hunk 4]
```diff
@@ -39,6 +40,9 @@ export const createApp = (data?: Partial<ApplicationData>,
         const session: Session | undefined = data ? configureSession(createSession(cookieSecret)) : undefined;
         session?.setExtraData(APPLICATION_DATA_KEY, data);
 
+        // @ts-ignore
+        session?.data.signin_info.company_number = 'NI000000';
```
Comment (geek-soft) [inline test/ApplicationFactory.ts:44]: This is needed for certain cases only, why is that added here? I am thinking having this here make is buried so deep that it will be difficult to understand how company number is set. I'd prefer to see it set explicitly for tests that need it.

File: src/models/CompanyAuthConfig.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+    oath_scope_prefix: string;
+    accountUrl: string;
```
Comment (geek-soft) [inline src/models/CompanyAuthConfig.ts:3]: Do we need separate ACCOUNT URL and CHS URL? What isCHS URL needed for?

Comment (akaJSaunders) [reply src/models/CompanyAuthConfig.ts:3]: In the enviroment variables there are these two:
```
ACCOUNT_URL=http://account.chs.local
ACCOUNT_WEB_URL=http://chs.local
```
ACCOUNT_WEB_URL was already present and is the CHS_URL in my code. I have tried using ACCOUNT_WEB for both of those variables, but it broke the authorisation journey.

Comment (geek-soft) [reply src/models/CompanyAuthConfig.ts:3]: I need to understand more what CHS_URL is needed for.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import { JWE, JWK } from 'node-jose';
+
+import CompanyAuthConfig from 'app/models/CompanyAuthConfig';
+
+interface AuthPayload {
+    nonce: string;
+    content: string;
+}
+
+export default class JwtEncryptionService {
```
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:13]: Again, no default exports please.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public constructor(private companyAuthConfig: CompanyAuthConfig) {}
+
+    public generateNonce(): string {
+        const bytes = randomBytes(5);
+        const buffer = Buffer.from(bytes);
+        return buffer.toString('base64');
+    }
+
+    public async jweEncodeWithNonce(returnUri: string, nonce: string): Promise<string> {
```
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:22]: Arg should not be called `returnUri` because service has nothing to do with URI's or redirects.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+import CompanyAuthConfig from 'app/models/CompanyAuthConfig';
+
+interface AuthPayload {
+    nonce: string;
+    content: string;
+}
+
+export default class JwtEncryptionService {
+    public constructor(private companyAuthConfig: CompanyAuthConfig) {}
```
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:14]: Config has 5 properties while only one is used, don't you think it is bad design? I would think it is better to keep that class generic as JWE encryption service which takes payload and key and returns encrypted payload.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts (hunk 7 above)
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:22]: I would consider rename as for me it is not encoding process, it is encrypting process.

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    const service: JwtEncryptionService = new JwtEncryptionService({
+        oath_scope_prefix: '',
+        chsUrl: '',
+        accountClientId: '',
+        accountRequestKey: 'pXf+qkU6P6SAoY2lKW0FtKMS4PylaNA3pY2sUQxNFDk=',
+        accountUrl: ''
+    });
+
+    it('should generate random nonce value in base64', () => {
```
Comment (geek-soft) [inline test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:16]: Not sure what that test does? It does not test randomness (because it is invoked only once). It does not test base 64 encoding really (because broken base 64 encoded string can still conform regex).

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        const nonce = service.generateNonce();
+        const test = /[A-Za-z0-9+/=]/.test(nonce);
+        assert.equal(nonce[nonce.length - 1], '=');
+        assert.equal(test, true);
+    });
+
+    it('should create an encrypted state string using nonce value', async () => {
+        const nonce = '2dsa=';
+        const state = await service.jweEncodeWithNonce('http://example.com', nonce);
+        const test = /eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2Iiwia2lkIjoia2V5In0..*/.test(state);
```
Comment (geek-soft) [inline test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:26]: What `eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2Iiwia2lkIjoia2V5In0` is? Would't it better to decrypt it back and compare decrypted value?

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+import JwtEncryptionService from 'app/modules/jwt-encryption-service/JwtEncryptionService';
+
+
+describe('JwtEncryptionService', () => {
+
+    const service: JwtEncryptionService = new JwtEncryptionService({
+        oath_scope_prefix: '',
+        chsUrl: '',
+        accountClientId: '',
+        accountRequestKey: 'pXf+qkU6P6SAoY2lKW0FtKMS4PylaNA3pY2sUQxNFDk=',
```
Comment (geek-soft) [inline test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:12]: That only confirms bad design I highlighted earlier.

File: src/models/CompanyAuthConfig.ts (hunk 5 above)
Comment (akaJSaunders) [reply src/models/CompanyAuthConfig.ts:3]: this is where they are both used:

```
return `${this.authConfig.accountUrl}/oauth2/authorise`.concat(
            '?',
            `client_id=${this.authConfig.accountClientId}`,
            `&redirect_uri=${this.authConfig.chsUrl}/oauth2/user/callback`,
            `&response_type=code`,
            `&scope=${scope}`,
            `&state=${encodedNonce}`);
```
accountURL serves as the domain of the URI we send the user to for the authorisation journey, and chsUrl is used for the callback parameter

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 9 above)
Comment (akaJSaunders) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:10]: How could this be tested? Running it multiple times and checking if values are different? two random values could happen to be equal (even tho chances are slim) and fail the test.

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 10 above)
Comment (akaJSaunders) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:21]: That would involve implementing a decode functionality for a test. What do you think?

File: test/ApplicationFactory.ts (hunk 4 above)
Comment (akaJSaunders) [reply test/ApplicationFactory.ts:44]: I originally wanted to add it to the `createSession` of `SessionFactory.ts`, where all other information is added however that is type checked. The current location is just after the `createSession` is called.

I cant figure out how to get access to the session through the individual tests, as the session is enclosed in the `app` that is created by this factory.

File: src/ContainerFactory.ts (hunk 1 above)
Comment (akaJSaunders) [reply src/ContainerFactory.ts]: I have created the class and used it for both the base controller and the container factory

File: src/middleware/CompanyAuthMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        try {
+            const uri = await this.getAuthRedirectUri(req, res, companyNumber);
+
+            loggerInstance().debug(`CompanyAuthMiddleware: Redirecting to ${uri}`);
+            return res.redirect(uri);
+
+        } catch (err){
+            console.log(err);
```
Comment (geek-soft) [inline src/middleware/CompanyAuthMiddleware.ts:55]: Remove, we do not log to console. It was only useful for debugging.

File: src/middleware/CompanyAuthMiddleware.ts [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            next(err);
+        }
+    }
+
+    async getAuthRedirectUri(req: Request, res: Response, companyNumber: string): Promise<string> {
+
+        const originalUrl: string = req.originalUrl;
+        const scope: string = this.authConfig.oathScopePrefix + companyNumber;
```
Comment (geek-soft) [inline src/middleware/CompanyAuthMiddleware.ts:63]: How is scope verified in the process? Does it matter? What would have happened if value provided by us was wrong?

File: test/ApplicationFactory.ts (hunk 4 above)
Comment (geek-soft) [reply test/ApplicationFactory.ts:44]: What test needs that?

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 9 above)
Comment (geek-soft) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:10]: Can randomness be increased beyond 5 bytes? That touches on question that Rhys had.

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 10 above)
Comment (geek-soft) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:21]: I do not see anything wrong in implementing decrypt function for the test especially that most of that heavy lifting is delegated to libraries. Feel better than the test that has magic number like `eyJhbGciO...`

These tests here look more like tests that needed to be added for sake of having them. Tests should be useful not just exist.

File: test/middleware/CompanyAuthMiddleware.test.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    };
+
+    const sessionStoreForAuthConfig = {
+        sessionCookieName: '__SID',
+        sessionCookieDomain: 'rebel1.aws.chdev.org',
+        sessionCookieSecureFlag: 'true',
+        sessionTimeToLiveInSeconds: 3600
+    };
+
+    const featureFlag: string = '1';
```
Comment (geek-soft) [inline test/middleware/CompanyAuthMiddleware.test.ts:51]: There is no test for value of 0. I'd think that having test for that making sure middleware is not triggered when feature is enabled is quite important.

File: .env.vagrant [hunk 15]
```diff
@@ -29,9 +29,14 @@ CDN_HOST=chs-dev.internal:8080
 
 APPEALS_API_URL=http://api.chs-dev.internal
 
+OATH_SCOPE_PREFIX=https://api.companieshouse.gov.uk/company/
 OAUTH2_TOKEN_URI=http://web.chs-dev.internal:4000/oauth2/token
 OAUTH2_CLIENT_ID=
 OAUTH2_CLIENT_SECRET=
+OAUTH2_REQUEST_KEY=
```
Comment (geek-soft) [inline .env.vagrant:36]: Are OAUTH2_REQUEST_KEY, OATH_SCOPE_PREFIX and ACCOUNT_URL still needed here? I'd suggest to remove these unless needed otherwise they just cause confusion that you experienced last time.

File: test/middleware/CompanyAuthMiddleware.test.ts [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+        );
+
+        await companyAuthMiddleware.handler(request, response, nextFunction);
+        nextFunction.didNotReceive();
+        response.received(1).redirect(Arg.any());
+
+    });
+});
+
+const createEncryptionService = (method: 'resolves' | 'rejects', encoded?: string | any)
```
Comment (geek-soft) [inline test/middleware/CompanyAuthMiddleware.test.ts:130]: Not a big fan of functions like that. Feels strange to have to say resolve / reject as via string arg. Here is how we used to do it before:

```
const emailService = createSubstituteOf<EmailService>(service => {
    service.send(Arg.any()).returns(Promise.reject(Error('Unexpected error')));
});
```

In above example it is clear to me what substitute of email service is setup to do.

File: test/middleware/CompanyAuthMiddleware.test.ts [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        );
+
+        const nextFunction = createSubstituteOf<NextFunction>();
+        const response = createSubstituteOf<Response>();
+        const request = { session: undefined } as Request;
+
+        try{
+            await companyAuthMiddleware.handler(request, response, nextFunction);
+        } catch (err) {
+            expect(err.message).to.equal(SESSION_NOT_FOUND_ERROR.message);
```
Comment (geek-soft) [inline test/middleware/CompanyAuthMiddleware.test.ts:73]: How does it work? Test says that next should be called while exception is being caught. I thought you will do verification of what arg next function was invoked with.

File: test/middleware/CompanyAuthMiddleware.test.ts [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+            sessionStore,
+            encryptionService,
+            companyAuthConfig,
+            sessionStoreForAuthConfig,
+            featureFlag
+        );
+
+        await companyAuthMiddleware.handler(request, response, nextFunction);
+        nextFunction.didNotReceive();
+        response.received(1).redirect(Arg.any());
```
Comment (geek-soft) [inline test/middleware/CompanyAuthMiddleware.test.ts:125]: Is making any redirect a purpose of the company middleware? Saying middleware works fine when any redirect is made is over simplification. Don't you think?

File: test/ApplicationFactory.ts (hunk 4 above)
Comment (akaJSaunders) [reply test/ApplicationFactory.ts:44]: Review Penalty Controller, the page that the middleware intercepts at the moment, if the `session?.data.signin_info.company_number` is not set, the middleware will redirect rather than let the page render

File: src/models/SessionConfig.ts [hunk 19]
```diff
@@ -0,0 +1,22 @@
+import { getEnvOrDefault, getEnvOrThrow } from 'app/utils/EnvironmentUtils';
+
+export interface SessionStoreConfig {
+    sessionCookieName: string;
+    sessionCookieDomain: string;
+    sessionCookieSecureFlag: string;
+    sessionTimeToLiveInSeconds: number;
+}
+
+export class SessionConfig {
```
[the remark this answers is not in the document]
Comment (akaJSaunders) [reply src/models/SessionConfig.ts:10]: tslint expects `createFromEnviromentVariables()` to have a type definition, and the interface provided that. What if I change the naming of the interface and remove the export? what do you think?

File: src/middleware/CompanyAuthMiddleware.ts (hunk 13 above)
Comment (akaJSaunders) [reply src/middleware/CompanyAuthMiddleware.ts:63]: This documentation explains the role of scope and their values quite well: https://companieshouse.atlassian.net/wiki/spaces/Arch/pages/1334771713/API+Filing+Service+Designs+3rd+Parties+and+CH

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 9 above)
Comment (akaJSaunders) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:10]: I increased it to 10, and it seems to work fine. Cant find documentation on any requirements.

File: src/middleware/CompanyAuthMiddleware.ts (hunk 13 above)
Comment (akaJSaunders) [reply src/middleware/CompanyAuthMiddleware.ts:61]: The convention for OAuth scopes is to provide the full URI of the resource including the domain then, add a period followed by the permission level e.g. 

the scope “https://api.companieshouse.gov.uk/user/profile.read” would provide the requesting client access to READ the user profile resource.

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts (hunk 9 above)
Comment (geek-soft) [reply test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:10]: Seems like nounce does not have to be too long, sometimes 32 bits (4 bytes) is enough and to be fairs as long as these are cryptographically random chance of collision is very small. Lets keep it at current length and lets generate it twice to make sure they are not the same.

File: src/models/SessionConfig.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/models/SessionConfig.ts:10]: Class provides type definitions as well. Why not to have a class with fields and static function?

File: .env.vagrant [hunk 20]
```diff
@@ -33,6 +33,8 @@ OAUTH2_TOKEN_URI=http://web.chs-dev.internal:4000/oauth2/token
 OAUTH2_CLIENT_ID=
 OAUTH2_CLIENT_SECRET=
 
+COMPANY_AUTH_FEATURE_FLAG=0
```
Comment (geek-soft) [inline .env.vagrant:36]: May I suggest different pattern? Out API used `CHIPS_FEATURE_ENABLED` which is bit more readable to me. Why not to have `COMPANY_AUTH_FEATURE_ENABLED` or `COMPANY_AUTH_VERIFICATION_FEATURE_ENABLED`

Names should be as clear as possible cause they will be maintained by people outside of the project.

File: src/middleware/CompanyAuthMiddleware.ts [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+                private readonly encryptionService: JwtEncryptionService,
+                private readonly authConfig: CompanyAuthConfig,
+                private readonly sessionStoreConfig: SessionStoreConfig,
+                private readonly featureFlag: boolean) {
+        super();
+    }
+
+    public handler: RequestHandler = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+
+        if(!this.featureFlag){
```
Comment (geek-soft) [inline src/middleware/CompanyAuthMiddleware.ts:28]: Reformat please.

File: src/middleware/CompanyAuthMiddleware.ts (hunk 13 above)
Comment (geek-soft) [reply src/middleware/CompanyAuthMiddleware.ts:61]: What happens if you provide bad value, e.g. `foo bar` as a scope?

File: test/middleware/CompanyAuthMiddleware.test.ts [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+    };
+
+    const sessionStoreForAuthConfig = {
+        sessionCookieName: '__SID',
+        sessionCookieDomain: 'rebel1.aws.chdev.org',
+        sessionCookieSecureFlag: 'true',
+        sessionTimeToLiveInSeconds: 3600
+    };
+
+    const redirectUrl: string = 'MOCK/oauth2/authorise?client_id=MOCK&redirect_uri=MOCK/oauth2/user/callback&response_type=code&scope=MOCKSC123123&state=';
```
Comment (geek-soft) [inline test/middleware/CompanyAuthMiddleware.test.ts:51]: Because you use MOCL for all elements of configuration it won't detect cases where you implemented bad string concatenation and for example `accountUrl` is used twice while `chsUrl` is never used.

Provide unique sample values for tests that will give us a chance to detect potential code issue.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+        return await JWE.createEncrypt({
+            format: 'compact',
+            fields: {
+                alg: 'dir',
+                enc: 'A128CBC-HS256'
+            }
+        }, key).update(payload).final();
+    }
+
+    public async jweDecryptWithNonce(content: string, requestKey: string): Promise<JWE.DecryptResult> {
```
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:38]: That name does not make sense. There is no nounce. Just base64 encoded value is being decrypted with key.

File: src/modules/jwt-encryption-service/JwtEncryptionService.ts [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+
+export class JwtEncryptionService {
+
+    public generateNonce(): string {
+        const bytes = randomBytes(10);
+        const buffer = Buffer.from(bytes);
+        return buffer.toString('base64');
+    }
+
+    public async jweEncryptWithNonce(content: string, nonce: string, requestKey: string): Promise<string> {
```
Comment (geek-soft) [inline src/modules/jwt-encryption-service/JwtEncryptionService.ts:14]: For me it makes more sense to combine `content: string, nonce: string` into one argument of object type. It gives more flexibility for it to be used in more generic use cases. Nounce is not relevant here, it is just payload.

After removing nounce from arg list please also remove `WithNounce` from name.

I'd generally keep it `encrypt` and `decrypt` cause class name defines what encryption it is.

File: src/middleware/CompanyAuthMiddleware.ts (hunk 13 above)
Comment (akaJSaunders) [reply src/middleware/CompanyAuthMiddleware.ts:61]: The scope is the concatenation of  OATH_SCOPE_PREFIX and companyNumber
If you set the environment variable OATH_SCOPE_PREFIX to an incorrect value, it will break when the user is intercepted by the middleware

File: test/modules/jwt-encryption-service/JwtEncryptionService.test.ts [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+        assert.equal(nonce1[nonce1.length - 1], '=');
+        assert.equal(nonce2[nonce2.length - 1], '=');
+        assert.equal(test1, true);
+        assert.equal(test2, true);
+    });
+
+    it('should create an encrypted state and decrypt correctly', async () => {
+        const nonce = '2dsa=';
+        const content = 'http://example.com';
+        const requestKey = 'pXf+qkU6P6SAoY2lKW0FtKMS4PylaNA3pY2sUQxNFDk=';
```
Comment (geek-soft) [inline test/modules/jwt-encryption-service/JwtEncryptionService.test.ts:28]: Can you use other key than one used in our env? I am not sure how it is being used and I'd rather avoid leaking secret that might be sensitive.
